lua/lspconfig/ui/prompt.lua

```diff
@@ -0,0 +1,54 @@
+local util = require('lspconfig.util')
+
+M = {}
+
+M.directory_picker = function(parent_dir)
+  return vim.fn.input("Root Directory: ", parent_dir, 'dir')
+end
+
+M.root_directory_prompt = function(filepath, border)
+    local filename = vim.fn.fnamemodify(filepath, ':t')
+    local parent_dir = util.path.dirname(filepath)
+
+    local bufnr = vim.api.nvim_create_buf(false, true)
+    local winnr = vim.api.nvim_open_win(bufnr, true, {relative='win', style='minimal', row=vim.o.lines, col=0, width=vim.o.columns, height=8, border=border})
+    vim.cmd(string.format("autocmd BufLeave <buffer=%d> ++once lua vim.api.nvim_win_close(%s, true)", bufnr, winnr))
+
+    local no_root_detected_message = string.format(" No root pattern matched for %s", filename)
+    local parent_dir_option = string.format(" p -> Set parent directory as project root %s", parent_dir)
+    local interactive_option = string.format(" c -> Interactively choose the project root")
+    local ignore_file_option = string.format(" i -> Ignore this file.")
+
+    vim.api.nvim_buf_set_lines(bufnr, 0, -1, true, {'', no_root_detected_message, '', parent_dir_option, interactive_option, ignore_file_option})
+    vim.api.nvim_buf_set_option(bufnr, 'modifiable', false)
+    vim.cmd [[redraw]]
+
+    local choices = {
+	["p"]=function() return parent_dir end,
+	["c"]=M.directory_picker,
+	["i"]=function() return nil end,
+    }
+
+    local user_input
+    local user_choice
+    local root_directory
+
+    while true do
```

Comment:
@folke Do you have a better solution for blocking input? I need to return a result from this function so I can't just map the callbacks and be done with it.

Reply:
That's pretty much what I do in WhichKey too.

The only thing I would change is:

```lua
  local ok, n = pcall(vim.fn.getchar)
 if not ok then break end
```

That properly handles keyboard interrupts too (user pressing `<control>c`)